src/core/cgroup.c

```diff
@@ -774,13 +774,70 @@ void cgroup_oomd_xattr_apply(Unit *u, const char *cgroup_path) {
                 unit_remove_xattr_graceful(u, cgroup_path, "user.oomd_omit");
 }
 
+int cgroup_log_xattr_apply(Unit *u, const char *cgroup_path) {
+        ExecContext *c;
+        size_t len;
+        size_t allowed_patterns_len = 0;
+        size_t denied_patterns_len = 0;
+        _cleanup_free_ char *patterns = NULL;
+        _cleanup_free_ char *allowed_patterns = NULL;
+        _cleanup_free_ char *denied_patterns = NULL;
+        int r;
+
+        assert(u);
+
+        c = unit_get_exec_context(u);
+        if (!c)
+                /* Some unit type have a cgroup context but no exec context, so we do not log
+                 * any error here to avoid confusion. */
+                return 0;
+
+        if (set_isempty(c->log_filter_allowed_patterns) && set_isempty(c->log_filter_denied_patterns)) {
+                unit_remove_xattr_graceful(u, cgroup_path, "user.journald_log_filter_patterns");
+                return 0;
+        }
+
+        r = set_make_nulstr(c->log_filter_allowed_patterns, &allowed_patterns, &allowed_patterns_len);
+        if (r < 0)
+                return log_debug_errno(r, "Failed to get nulstr from set: %m");
+
+        r = set_make_nulstr(c->log_filter_denied_patterns, &denied_patterns, &denied_patterns_len);
+        if (r < 0)
+                return log_debug_errno(r, "Failed to get nulstr from set: %m");
+
+        /* Compute actual length for nulstrs and extended attribute. set_make_nulstr() behaves the same way
+         * as strv_make_nulstr():
+         * - If 'len == 0': we allocated an empty nulstr (1 byte) containing '\0', meaning size is actually 1.
+         * - If 'len > 0': we allocated space to store the strings but we do not account for the
+         *   nul-termination character ('\0') of the last string, nor the nul-termination character of the
+         *   nulstr, meaning size is actually 'len + 2'.
+         * This is necessary as we need to copy the exact data returned by set_make_nulstr() in order for
+         * systemd-journald to parse it properly. */
+        allowed_patterns_len = allowed_patterns_len ? allowed_patterns_len + 2 : 1;
+        denied_patterns_len = denied_patterns_len ? denied_patterns_len + 2 : 1;
+        len = allowed_patterns_len + 1 + denied_patterns_len;
+
+        patterns = new(char, len);
+        if (!patterns)
+                return log_oom_debug();
+
+        memcpy_safe(patterns, allowed_patterns, allowed_patterns_len);
+        memcpy_safe(&patterns[allowed_patterns_len], "\xff", 1);
```

Comment:
Do we even need this separator at all if we're encoding the full nulstr in the xattr? You could just search for the double nul terminator of the first nulstr to find the second nulstr.

```
denylist = mempmem_safe(xattr, xattr_size, "\0\0", 2);
```

Reply:
I prefer to keep the delimitation character. This way it's clear where the split is.

Also, what about empty strings? For example, `\0\0foo\0...`: is it `\0\0foo\xff\0...` or `\0\0\xfffoo\0...`?